### PR TITLE
Colorbar extend and norm 

### LINKED ISSUE #268: Colorbar `extend` and 'norm' parameters are deprecated since 3.3 and will be removed two minor releases later
`matplotlib` will no longer be using the `extend` and `norm` parameters for colorbars as stated [here](https://matplotlib.org/3.3.1/api/api_changes.html#effectless-parameters-of-figure-colorbar-and-matplotlib-colorbar-colorbar). We will need to address this in the following examples in the gallery. This list is current at the time of opening this issue and may need to be updated as other examples are added.

Examples that use `extend`:
[CB_Height.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Colors/CB_Height.html#sphx-glr-gallery-colors-cb-height-py)
[CB_Rain.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Colors/CB_Rain.html#sphx-glr-gallery-colors-cb-rain-py)
[CB_Temperature.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Colors/CB_Temperature.html#sphx-glr-gallery-colors-cb-temperature-py)
[NCL_lcnative_1_lg.py](https://geocat-examples.readthedocs.io/en/latest/gallery/MapProjections/NCL_lcnative_1_lg.html#sphx-glr-gallery-mapprojections-ncl-lcnative-1-lg-py)
[NCL_native_2.py](https://geocat-examples.readthedocs.io/en/latest/gallery/MapProjections/NCL_native_2.html#sphx-glr-gallery-mapprojections-ncl-native-2-py)
[NCL_proj_1.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Contours/NCL_proj_1.html)
[NCL_proj_3.py](https://geocat-examples.readthedocs.io/en/latest/gallery/MapProjections/NCL_proj_3_lg.html#sphx-glr-gallery-mapprojections-ncl-proj-3-lg-py)
[NCL_sat_3.py](https://geocat-examples.readthedocs.io/en/latest/gallery/MapProjections/NCL_sat_3.html#sphx-glr-gallery-mapprojections-ncl-sat-3-py)
[NCL_vector_1.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Vectors/NCL_vector_1.html)

Examples that use `norm`:
[NCL_overlay_6.py](https://geocat-examples.readthedocs.io/en/latest/gallery/Overlays/NCL_overlay_6.html)

### PR DESCRIPTION
**Pull request by michaelavs**
Updating plots due to matplotlib version depreciations of `extend` and `norm`
closes #268

### CODE REVIEW

File: Plots/Overlays/NCL_overlay_6.py [hunk 1]
```diff
@@ -131,19 +130,19 @@
...
 cax1.tick_params(size=0)
 
-#
+###############################################################################
 # Overlay streamlines
-#
+
 with np.errstate(
-        invalid='ignore'
-):  # Indeed not needed, just to get rid of warnings about numpy's NaN comparisons
```
Comment (hCraker) [inline Plots/Overlays/NCL_overlay_6.py:146]: Why did you remove this comment? It is necessary to explain to novice programmers why we are choosing to ignore warnings. This PR is just to fix issues with the `extend` and `norm` arguments for colorbars, so let's fix other things in a separate PR.

File: Plots/Overlays/NCL_overlay_6.py [hunk 2]
```diff
@@ -199,8 +201,7 @@
 
 # Draw wind vectors
 with np.errstate(
-        invalid='ignore'
-):  # Indeed not needed, just to get rid of warnings about numpy's NaN comparisons
```
Comment (hCraker) [inline Plots/Overlays/NCL_overlay_6.py:203]: Same as above. Let's focus on just the `extend` and `norm` issues in this PR.

File: Plots/Overlays/NCL_overlay_6.py [hunk 3]
```diff
@@ -209,13 +210,15 @@
                   transform=ccrs.PlateCarree(),
                   headwidth=5,
                   cmap=wind_cmap,
-                  norm=norm,
+                  # norm=norm,
```
Comment (hCraker) [inline Plots/Overlays/NCL_overlay_6.py:213]: Are you planning on leaving this line in as a comment, or do you plan to remove it all together? It's a bit ambiguous.

File: Plots/Overlays/NCL_overlay_6.py [hunk 4]
```diff
@@ -209,13 +210,15 @@
...
                   zorder=4)
+# Add color bar
 plt.colorbar(Q,
              cax=cax2,
              ticks=np.arange(-20, 110, 10),
-             norm=norm,
+             # norm=norm,
```
Comment (hCraker) [inline Plots/Overlays/NCL_overlay_6.py:219]: Same as above in line 213.

File: Plots/Colors/CB_Temperature.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 """
 CB_Temperature.py
-=================
+================
```
Comment (hCraker) [inline Plots/Colors/CB_Temperature.py:3]: Sphinx will throw a warning with this saying that the title underline is too short. Make sure the underline is the same length as the title.

File: Plots/MapProjections/NCL_lcnative_1.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 """
 NCL_lcnative_1.py
-=================
```
Comment (hCraker) [inline Plots/MapProjections/NCL_lcnative_1.py:3]: Sphinx will throw a warning with this saying that the title underline is too short. Make sure the underline is the same length as the title.

File: Plots/Overlays/NCL_overlay_6.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 """
 NCL_overlay_6.py
-================
+===============
```
Comment (hCraker) [inline Plots/Overlays/NCL_overlay_6.py:3]: Sphinx will throw a warning with this saying that the title underline is too short. Make sure the underline is the same length as the title.

File: Plots/Colors/CB_Height.py [hunk 8]
```diff
@@ -64,68 +64,67 @@
...
+        vmin=100,
+        vmax=1600,
+        cmap=newcmp,
+        add_colorbar=False)
+
+    # Add color bar
+    cbar_ticks = np.arange(100, 1600, 100)
+    cbar = plt.colorbar(hgt, orientation='vertical', shrink=0.8, pad=0.05, extendrect=True)
+    cbar.ax.tick_params(labelsize=10)
+    cbar.set_ticks(cbar_ticks)
```
Comment (hCraker) [inline Plots/Colors/CB_Height.py:91]: The colorbar ticks can be set with `plt.colorbar()` by using the `ticks` argument. What you have works, but this way is more concise if you would like to use it instead.

**Review by hCraker — COMMENTED** [review]
I've left a few comments asking for clarification for some of the edits made. I noticed you made some formatting changes, small changes to figure titles, and changed comments in multiple files. To keep ourselves organized, let's primarily focus on fixing the color bar issues with this PR. 

`NCL_overlay_6.py` is now generating incorrect figures on the documentation when run with `make html`.  This is what it looks like now. I remember `NCL_overlay_6.py` being a particularly difficult script to write with the multiple colorbars, so let me know if you run into trouble with it. I may be able to help.
image

This is a good start, but let's focus on one issue at a time especially since this is a multi-file PR. All of the other files look great when the documentation is built. I don't see any glaring issues with the figures excluding overlay_6.

File: Plots/MapProjections/NCL_proj_3.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
+
 """
-NCL_proj_3.py
-=============
+NCL_proj_3_lg.py
```
[the remark this answers is not in the document]
Comment (michaelavs) [reply Plots/MapProjections/NCL_proj_3.py:3]: I think my local version is from an older version of the code. I am currently going through and comparing things like this to make sure any additional edits are not detrimental (like in this one) and are formatted correctly. Most seem to be relatively quick and simple fixes, but I think I will remove this PR and reopen it with completely new pulls of the files from the repo just to completely ensure that there aren't any bigger differences that just haven't been caught yet, and I don't want to mess up any builds because of having a different version of a code locally for whatever reason. 

The new PR will most likely be published tomorrow AM, but I will work on getting everything re-done in preparation for that tonight.

[the remark this answers is not in the document]
Comment (michaelavs) [reply Plots/MapProjections/NCL_proj_3.py:3]: That was a slight lie, I'm not going to delete this PR. I have cleared out the versions of the codes that I have locally and pulled them from the master. I'm going to now re upload them to this PR, make sure the only differences are from the cbar edits and then go ahead with adding in the cbar changes again. You will probably get a couple of emails during this process, so sorry for the spam in advance!

**Review by anissa111 — COMMENTED** [review]
This looks really good! 

My only comment is that the surface temperature color bar seems to be slightly different. I've included the difference below. It seems like the ranges for the colors might have been altered slightly.

NCL original:
image

Previous python generated image:
image


New generated image:
image

**Comment by michaelavs** [comment]
> This looks really good!
> 
> My only comment is that the surface temperature color bar seems to be slightly different. I've included the difference below. It seems like the ranges for the colors might have been altered slightly.
> 
> NCL original:
> image
> 
> Previous python generated image:
> image
> 
> New generated image:
> image

I've been looking into this, and I'm not 100% sure the answer of why there is a difference. I did `np.nanmax` and `np.nanmin` on the `t` variable (which is what the colorbar should be derived from) and found that the max is about 87 degrees and the min is about -16 degrees. So it may make sense why the colorbar is not wanting to have an extent that is bigger or smaller than those two numbers, but it doesn't quite explain why this wasn't an issue in the initial version (both python and NCL).  I'll keep digging on this, but if you have any ideas on how to resolve this, let me know!

**Comment by michaelavs** [comment]
> This looks really good!
> 
> My only comment is that the surface temperature color bar seems to be slightly different. I've included the difference below. It seems like the ranges for the colors might have been altered slightly.
> 
> NCL original:
> image
> 
> Previous python generated image:
> image
> 
> New generated image:
> image

Ok this should be resolved now. The problem was from removing 'norm=norm` in the `ax.quiver` call. This function has not depreciated `norm=norm`, but the `plt.colorbar()` has, so it only needed to be deleted from there.

**Review by hCraker — COMMENTED** [review]
I went through all of the files you changes to ensure the figures look the same as before. I noticed a discrepancy in colorbar for `NCL_sat_3.py`. Once you change that you should be all good to go.
Last Python Version:
image

This PR:
image

**Comment by hCraker** [comment]
On `NCL_sat_3.py`, the value -1.5 is not on the very far left of the colorbar like it is in the original NCL examples. Once you make sure the colorbar starts with the right value, you should be good to go.
Original NCL:
image

Your current version in python:
image

**Review by hCraker — APPROVED** [review]
Thanks for fixing the color bar on the sat_3 example. Everything looks good to me!

**Review by anissa111 — APPROVED** [review]
This looks good, thank you for doing this!
